controllers/registry.go

```diff
@@ -213,53 +214,102 @@ func (r *DPAReconciler) ReconcileRegistries(log logr.Logger) (bool, error) {
+		// 	if err != nil {
+		// 		return err
+		// 	}
+		// 	// update the Registry Deployment template
+		// 	err = r.buildRegistryDeployment(registryDeployment, &bsl, &dpa)
+		// 	return err
+		// })
+
+		// if err != nil {
+		// 	return false, err
+		// }
+
+		// //TODO: Review registry deployment status and report errors and conditions
+
+		// if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+		// 	// Trigger event to indicate registry deployment was created or updated
+		// 	r.EventRecorder.Event(registryDeployment,
+		// 		corev1.EventTypeNormal,
+		// 		"RegistryDeploymentReconciled",
+		// 		fmt.Sprintf("performed %s on registry deployment %s/%s", op, registryDeployment.Namespace, registryDeployment.Name),
+		// 	)
+		// }
 
-			// Setting Registry Deployment selector if a new object is created as it is immutable
-			if registryDeployment.ObjectMeta.CreationTimestamp.IsZero() {
-				registryDeployment.Spec.Selector = &metav1.LabelSelector{
-					MatchLabels: map[string]string{
-						"component": registryName(&bsl),
-					},
-				}
-			}
+	}
 
-			err := controllerutil.SetControllerReference(&dpa, registryDeployment, r.Scheme)
-			if err != nil {
-				return err
-			}
-			// update the Registry Deployment template
-			err = r.buildRegistryDeployment(registryDeployment, &bsl, &dpa)
-			return err
-		})
+	return true, nil
+}
 
-		if err != nil {
-			return false, err
-		}
+func (r *DPAReconciler) BuildRegistryContainers(veleroContainer *corev1.Container) (containers []corev1.Container, err error) {
+	dpa := oadpv1alpha1.DataProtectionApplication{}
+	if err := r.Get(r.Context, r.NamespacedName, &dpa); err != nil {
+		return nil, err
+	}
+	if !dpa.BackupImages() {
+		return nil, nil
+	}
 
-		//TODO: Review registry deployment status and report errors and conditions
+	backupStorageLocationList := velerov1.BackupStorageLocationList{}
 
-		if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
-			// Trigger event to indicate registry deployment was created or updated
-			r.EventRecorder.Event(registryDeployment,
-				corev1.EventTypeNormal,
-				"RegistryDeploymentReconciled",
-				fmt.Sprintf("performed %s on registry deployment %s/%s", op, registryDeployment.Namespace, registryDeployment.Name),
-			)
+	// Fetch the configured backupstoragelocations
+	if err := r.List(r.Context, &backupStorageLocationList); err != nil {
+		return nil, err
+	}
+	// sorting items by creationTimeStamp from oldest to newest
+	sort.Slice(backupStorageLocationList.Items, func(i, j int) bool {
+		return backupStorageLocationList.Items[i].CreationTimestamp.Before(&backupStorageLocationList.Items[j].CreationTimestamp)
+	})
+	// Loop through all the configured BSLs and create registry for each of them
+	for i, bsl := range backupStorageLocationList.Items {
+		// debug server listens on port +1 so increment by 2
+		containerPort := int32(5000+i*2)
```

Comment:
Is there some way to get a port that is open rather than hardcoding the ports?

Reply:
probably.

Reply:
After some thinking I don’t think there is a way to dynamically at velero deployment building time to know ahead of time what ports will be open since it depends on what other containers are binding which isn’t immediately visible by scanning deployment spec alone. Just have to test hardcoded values aren’t conflicting with the some other containers in the pod.

Reply:
It is possible velero image updates start binding to more ports (without exposing in corev1.Container.Ports) in the future and block our access. This isn't visible in deployment spec.

Reply:
alternative architecture that might allow for open port detection instead of hardcoding is to start registry process inside velero container itself from the plugin.

Reply:
Maybe open port can be discovered when container starts but then Exec liveness probe might have to wait for environment variable to be set. registry container will also have to update ENV in the velero container or alternatively both reading/writing port value to a shared volumeMount.